flocker/testtools/_flaky.py

```diff
@@ -22,10 +37,179 @@ def flaky(jira_key):
+    min_passes = field(int, mandatory=True,
+                       invariant=lambda x: (x > 0, "must pass at least once"))
+    jira_keys = pset_field(unicode, optional=False)
+
+    __invariant__ = lambda x: (
+        x.max_runs >= x.min_passes and len(x.jira_keys) > 0,
+        "Can't pass more than we run and must provide a jira key")
+
+    def to_dict(self):
+        return {
+            'max_runs': self.max_runs,
+            'min_passes': self.min_passes,
+            'jira_keys': set(self.jira_keys),
+        }
+
+
+def retry_flaky(run_test_factory=None):
+    """
+    Wrap a ``RunTest`` object so that flaky tests are retried.
+    """
+    if run_test_factory is None:
+        run_test_factory = testtools.RunTest
+
+    return partial(_RetryFlaky, run_test_factory)
+
+
+class _RetryFlaky(testtools.RunTest):
+    """
+    ``RunTest`` implementation that retries tests that fail.
+    """
+    # XXX: This should probably become a part of testtools:
+    # https://bugs.launchpad.net/testtools/+bug/1515933
+
+    def __init__(self, run_test_factory, case, *args, **kwargs):
+        self._run_test_factory = run_test_factory
+        self._case = case
+        self._args = args
+        self._kwargs = kwargs
+
+    def _run_prepared_result(self, result):
+        """
+        Run the test with a result that conforms to testtools' extended
+        ``TestResult`` interface.
+
+        This overrides a method in base ``RunTest`` which is intended to be
+        overwritten.
+        """
+        flaky = _get_flaky_annotation(self._case)
+        if flaky is not None:
+            return self._run_flaky_test(self._case, result, flaky)
+
+        # No flaky attributes? Then run as normal.
+        return self._run_test(self._case, result)
+
+    def _run_test(self, case, result):
+        """
+        Run ``case`` with the ``RunTest`` we are wrapping.
+
+        :param testtools.TestCase case: The test to run.
+        :param testtools.TestResult result: The test result to report to.
+            Must conform to testtools extended test result interface.
+        :return: The modified ``result``.
+        """
+        run_test = self._run_test_factory(case, *self._args, **self._kwargs)
+        return run_test._run_prepared_result(result)
+
+    def _run_flaky_test(self, case, result, flaky):
+        """
+        Run a test that has been decorated with the `@flaky` decorator.
+
+        :param TestCase case: A ``testtools.TestCase`` to run.
+        :param TestResult result: A ``TestResult`` object that conforms to the
+            testtools extended result interface.
+        :param _FlakyAnnotation flaky: A description of the conditions of
+            flakiness.
+
+        :return: A ``TestResult`` with the result of running the flaky test.
+        """
+        successes = 0
+        results = []
+
+        while successes < flaky.min_passes and len(results) < flaky.max_runs:
```

Comment:
:grey_question: Optional: change the second half of the condition to: (len(results) + flaky.min_passes - successes <= flaky.max_runs)
It'll save us some runs/time/money/energy if a flaky test configured for 5/2 has already failed 4 times.